adr/context.py

```diff
@@ -0,0 +1,153 @@
+      }]
+]
+"""
+These are commonly used arguments which can be re-used. They are shared to
+provide a consistent CLI across recipes/queries.
+"""
+
+
+def extract_context_names(query):
+    """
+    Extract contexts list from query object
+    This is to mimic json-e render() function
+    Given using only $eval in query
+    :param query: query object to extract context
+    :return: list of context names
+    """
+    if isinstance(query, dict):
+        for k, v in query.items():
+            if k == "$eval":
+                yield v
+            else:
+                yield from extract_context_names(v)
+    elif isinstance(query, list):
+        for v in query:
+            yield from extract_context_names(v)
+
+
+def _get_destination(definition):
+    """
+    Extract destination name from definition
+    :param list definition: definition of context, in form of list or str
+    :return:
+    """
+    if not isinstance(definition, list):
+        return None
+    if len(definition) < 2:
+        raise TypeError("Invalid definition {}".format(definition))
+    arg_def = definition[1]
+    keys = definition[0]
+    # If dest is defined, use it
+    if 'dest' in arg_def:
+        return arg_def['dest']
+    # Otherwise use the long form of arg key name
+    val = next((key for key in keys if key.startswith('--')), None)
+    return val[2:] if val else None
+
+
+def _get_context_definition(context_name, definitions):
+    return next((definition for definition in definitions
+                 if _get_destination(definition) == context_name),
+                None)
+
+
+def get_context_definitions(definitions):
+    """
+    Build full context definitions from list of contexts with pre-defined definitions
+    :param list definitions: list of context definition, in which each definition is an array
+    :return:
+    """
+    result = []
+    destinations = set()
+    for name in definitions:
+        if isinstance(name, list):
+            definition = name
+        elif not isinstance(name, str):
+            raise TypeError("Only accept str/list, but found {}".format(type(name)))
+        else:
+            # Look at the custom definition first
+            definition = _get_context_definition(name, definitions)
+            # Not found? try COMMON_CONTEXTS
+            if not definition:
+                definition = _get_context_definition(name, COMMON_CONTEXTS)
+                # Still not found! bad definition
+                if not definition:
+                    raise AttributeError("Cannot find definition for context {}".format(name))
+        definition_dst = _get_destination(definition)
+        if definition_dst not in destinations:
+            result.append(definition)
+            destinations.add(definition_dst)
+
+    return result
+
+
+def merge_definitions(definitions, src_definitions):
+    """
+    Merge src to destination
+    :param definitions:
+    :param src_definitions:
```

Comment:
I have to admit I'm a little lost trying to follow these functions around. Could you expand your docstrings help to try and make it a bit more clear what each function is doing and what things like "src" and "destination" mean? I also find the term "definition" to be a bit vague.

Btw, the docs have the [napoleon](http://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) sphinx extension installed, so please try to follow the Google style guide for docstrings in there.

Reply:
I updated the docstring of this function. This function is used to merge/combine all query context definitions and/or run context definitions to recipe context definition. Currently, we have to merge all definitions into recipe context definition and parse only one time. Otherwise, the "--help" doesn't work well (I put TO-DO to mark this problem and will return to it later).